### PR TITLE
viper: assertions

### PR DESCRIPTION
**Pull request by ggreif**
This adds the various kinds of static assertions to the Motoko grammar.

TODO:
- [ ] tokens
  - [x] `invariant`
  - [ ] `static`
- [x] `> assert:;` -> extend `Error_reporting.terminal2token`
- [x] `AssertE` _how_ argument
- [ ] purity declaration (`function` instead of `method` in Viper)
- [x] pre/post-conditions (`assert:func`, `assert:return`)
- [ ] loop invariants (`assert:loop`, `assert:invariant`, `assert:break` for enter, stay and exit)
- [x] `implies` (translates to Viper `==>`)

### CODE REVIEW

**Comment by crusso** [comment]
Hm, I'm not sure this is the way to go though it might work for a quick hack.

In general, invariants belong to just loops, pre and post to just methods headers etc. Allowing them all over the place would allow  eg. preconditions to incorrectly refer to local variables etc.

File: src/mo_frontend/assertions.mly [hunk 1]
```diff
@@ -0,0 +1,11 @@
+%%
+
+%public exp_nondec(B) :
+  | ASSERT COLON STABLE e=exp_nest
```
Comment (ggreif) [inline src/mo_frontend/assertions.mly:4]: should be `INVARIANT`

**Review by ggreif — COMMENTED** [review]
Yes, this is just for a quick hack. `assert:invariant` may occur in `actor` at top-level.
`assert:func` and `assert:return` are ignored (print as comments).

**Comment by ggreif** [comment]
> preconditions to incorrectly refer to local variables

Right. This will ultimately cause hard errors on the Viper side.

File: src/mo_frontend/assertions.mly [hunk 2]
```diff
@@ -0,0 +1,23 @@
+%token INVARIANT
+(* I get
+> Error: 54 states have shift/reduce conflicts.
```
Comment (ggreif) [inline src/mo_frontend/assertions.mly:3]: To me this looks like a `menhir` bug.

File: src/viper/syntax.ml [hunk 3]
```diff
@@ -72,6 +72,8 @@ and stmt' =
   | IfS of exp * seqn * seqn
   | WhileS of exp * invariants * seqn
   | LabelS of id * invariants
+  | PreconditionS of exp
```
Comment (ggreif) [inline src/viper/syntax.ml:75]: These are currently get translated as `stmt` lists, but before constructing the `MethodI` they get converted to the pre/post entries of `MethodI`, and stripped from the body. 63189ba6b5b0d32fa95b3edfad89ff5eaad66881

File: test/viper/invariant.mo [hunk 4]
```diff
@@ -4,10 +4,13 @@ actor {
 
   var count = 0 : Int;
 
-  assert claimed and not (-1 == -1) and (-42 == -42) or true;
-  assert count > 0;
+  assert:invariant claimed and not (-1 == -1) and (-42 == -42) or true;
```
Comment (ggreif) [inline test/viper/invariant.mo:7]: Of course we can change this to just be `invariant`.

File: test/viper/invariant.mo [hunk 5]
```diff
@@ -4,10 +4,13 @@ actor {
...
+  assert:invariant count > 0;
+  assert:invariant not claimed implies count == 0;
 
   public shared func claim() : async () {
+      assert:func count >= 0;
+      assert:return count >= 0;
```
Comment (aterga) [inline test/viper/invariant.mo:13]: What do these two lines mean?

Comment (aterga) [reply test/viper/invariant.mo:13]: Is this how preconditions and postconditions being rendered?

Comment (ggreif) [reply test/viper/invariant.mo:13]: We don't have dedicated Motoko syntax yet, so they are just Motoko statements.

Comment (aterga) [reply test/viper/invariant.mo:13]: Sounds good. Thanks!

File: test/viper/invariant.mo (hunk 4 above)
Comment (aterga) [reply test/viper/invariant.mo:7]: This looks much better, thanks!

**Review by aterga — APPROVED** [review]

**Pull request merged**